src/video_core/renderer_vulkan/vk_pipeline_cache.cpp

```diff
@@ -0,0 +1,647 @@
+    case AttribType::Ubyte:
+        return sizeof(u8) * attrib.size;
+    }
+}
+
+vk::Format ToVkAttributeFormat(VertexAttribute attrib) {
+    switch (attrib.type) {
+    case AttribType::Float:
+        switch (attrib.size) {
+        case 1:
+            return vk::Format::eR32Sfloat;
+        case 2:
+            return vk::Format::eR32G32Sfloat;
+        case 3:
+            return vk::Format::eR32G32B32Sfloat;
+        case 4:
+            return vk::Format::eR32G32B32A32Sfloat;
+        }
+    case AttribType::Ubyte:
+        switch (attrib.size) {
+        case 4:
+            return vk::Format::eR8G8B8A8Uint;
+        default:
+            fmt::print("{}\n", attrib.size.Value());
+            UNREACHABLE();
+        }
+
+    default:
+        LOG_CRITICAL(Render_Vulkan, "Unimplemented vertex attribute type {}", attrib.type.Value());
+        UNREACHABLE();
+    }
+
+    return vk::Format::eR32Sfloat;
+}
+
+vk::ShaderStageFlagBits ToVkShaderStage(std::size_t index) {
+    switch (index) {
+    case 0:
+        return vk::ShaderStageFlagBits::eVertex;
+    case 1:
+        return vk::ShaderStageFlagBits::eFragment;
+    case 2:
+        return vk::ShaderStageFlagBits::eGeometry;
+    default:
+        LOG_CRITICAL(Render_Vulkan, "Invalid shader stage index!");
+        UNREACHABLE();
+    }
+
+    return vk::ShaderStageFlagBits::eVertex;
+}
+
+PipelineCache::PipelineCache(const Instance& instance, TaskScheduler& scheduler,
+                             RenderpassCache& renderpass_cache)
+    : instance{instance}, scheduler{scheduler}, renderpass_cache{renderpass_cache} {
+    descriptor_dirty.fill(true);
+
+    LoadDiskCache();
+    BuildLayout();
+    trivial_vertex_shader = Compile(GenerateTrivialVertexShader(), vk::ShaderStageFlagBits::eVertex,
+                                    instance.GetDevice(), ShaderOptimization::Debug);
+}
+
+PipelineCache::~PipelineCache() {
+    vk::Device device = instance.GetDevice();
+
+    SaveDiskCache();
+
+    device.destroyPipelineLayout(layout);
+    device.destroyPipelineCache(pipeline_cache);
+    device.destroyShaderModule(trivial_vertex_shader);
+    for (std::size_t i = 0; i < MAX_DESCRIPTOR_SETS; i++) {
+        device.destroyDescriptorSetLayout(descriptor_set_layouts[i]);
+        device.destroyDescriptorUpdateTemplate(update_templates[i]);
+    }
+
+    for (auto& [key, module] : programmable_vertex_shaders.shader_cache) {
+        device.destroyShaderModule(module);
+    }
+
+    for (auto& [key, module] : fixed_geometry_shaders.shaders) {
+        device.destroyShaderModule(module);
+    }
+
+    for (auto& [key, module] : fragment_shaders.shaders) {
+        device.destroyShaderModule(module);
+    }
+
+    for (const auto& [hash, pipeline] : graphics_pipelines) {
+        device.destroyPipeline(pipeline);
+    }
+
+    graphics_pipelines.clear();
+}
+
+void PipelineCache::BindPipeline(const PipelineInfo& info) {
+    ApplyDynamic(info);
+
+    u64 shader_hash = 0;
```

Comment:
```suggestion
    std::size_t shader_hash = 0;
```
macOS complains otherwise. Vide previous "fix hashCombine usage to make macos happy" commit